### DISCUSSION TITLE
Browser Push Notifications

### DISCUSSION THREAD

**Discussion by sjonez** [dgreif/ring#911] [Ideas]
I noticed on the web portal that you can access at ring.com it now has browser push notifications available for dings/motion. Is this something you can make use of so that dings are event driven rather than polled?

**Comment by sjonez** [comment]
I think that notifications (to chrome at least) use firebase messaging, so I don't know if something like this would help https://github.com/MatthieuLemoine/push-receiver

**Comment by dgreif** [comment]
Hey @sjonez, thanks for pointing this out!  I also noticed this feature a few weeks ago and did some digging into fcm to see if we could leverage it here.  I didn't find `push-receiver` in my research, but that definitely looks like a promising option.  I'm currently working on camera streaming in the `beta` branch, but will probably look into this once streaming is stable.  Would make a _huge_ difference if we could get push notifications instead of polling the API!

**Reply by dgreif** [reply]
I spent some time trying out `push-receiver` last night and unfortunately it doesn't appear to be working.  I have a custom FCM project set up that I tried subscribing it to and publishing a test message from, but the client never received the message.  Also tried hooking it up to Ring push notifications, and again I did not receive any messages.  It appears the `push-receiver` project hasn't been updated for several years, so it's possible that it simply doesn't work anymore.  Happy to try another solution if others have suggestions though!

**Reply by sjonez** [reply]
I did notice that there have been a few pull requests since the latest release, so might be worth incorporating the latest code changes... ?

Otherwise there's a quite recently updated python module called Pullkin that seems to do the same if that helps work out a solution? 

https://github.com/WhiteApfel/Pullkin

**Reply by sjonez** [reply]
Good news! I found this very recent fork of push-receiver: https://github.com/Eneris/push-receiver/

The code is quite different to the original, but I tested it with my firebase sender ID and the notification came straight through!

**Reply by dgreif** [reply]
Trying it out now.  How did you send the message from firebase?  Just using the "New Notification" function on the FCM dashboard?  I can't seem to get a notification to come through yet 😕

**Reply by sjonez** [reply]
Here's the code I used

```
const PushReceiver = require('@eneris/push-receiver').default
const parsedArgs = require('yargs/yargs')(process.argv.slice(2)).argv;

if (!parsedArgs.senderId) {
    console.error('Missing senderId')
    return
}

(async () => {
    const instance = new PushReceiver({
        logLevel: parsedArgs.logLevel || 'DEBUG',
        senderId: parsedArgs.senderId,
        persistentIds: [], // Recover stored ids of all previous notifications
    })

    const stopListeningToCredentials = instance.onCredentialsChanged(({ oldCredentials, newCredentials }) => {
        console.log('Client generated new credentials.', newCredentials)
        // Save them somewhere! And decide if thing are needed to re-subscribe
    })

    const stopListeningToNotifications = instance.onNotification(({ notification }) => {
        // Do someting with the notification
        console.log('Notification received', notification)
    })

    await instance.connect()

    if (parsedArgs.serverId) {
        await instance.testMessage(parsedArgs.serverId)
    }

    stopListeningToCredentials()
    stopListeningToNotifications()
    instance.destroy()
})()
```
I then started that up with `node test.js --senderId=<FIREBASE_SENDER_ID>` using the sender ID from my firebase dashboard

The debug output from that gave me a FCM token, which I entered on firebases "Send Test Message" function at https://console.firebase.google.com/project/PROJECT_NAME/notification/compose

That resulted in:

```
[PUSH_RECEIVER_DEBUG] emit ON_MESSAGE_RECEIVED {
  message: {
    data: {
      'gcm.n.e': '1',
      'google.c.a.ts': 'xxxx',
      'google.c.a.udt': '0',
      'google.c.a.e': '1',
      'google.c.a.c_id': 'xxxx',
      'google.c.a.c_l': 'test'
    },
    from: '21193289806',
    priority: 'high',
    notification: {
      title: 'test message title',
      body: 'test message body',
      tag: 'campaign_collapse_key_xxxx'
    },
    fcmMessageId: '686d2657-7b67-471d-bc0c-xxxxx',
    collapse_key: 'campaign_collapse_key_34540972822xxxx'
  },
  persistentId: '0:1648831895536089%7031b2e6xxxxx'
}
```

**Reply by dgreif** [reply]
Ahh, the "Send test message" piece is what I had been missing!  That worked for me 🎉 🚀.  Now that I know this piece is functional, I'll keep playing around with the Ring side to see if I can get messages coming through.  Thank you for persisting on this, it's one of the pieces of the repo I've been wanting to improve for years.  Getting close!

**Reply by sjonez** [reply]
Yep, that's the bit I'm struggling with - how to make use of that with Ring. I assume we need to register our hardware ID with rings own notification endpoint, while also subscribing on the firebase side of things... just not sure how they connect together.

Monitoring browser traffic while enabling notifications might make things clearer...

**Reply by dgreif** [reply]
In the browser, they use the `device` endpoint with a PATCH to set the `push_notification_token`.  Just override my `device` from the browser and was able to receive the messages via `push-receiver`!!!!  Still need to get the PATCH working from the node side, but at least we know for certain the notifications can flow from Ring to the nodejs side 🎉

**Reply by sjonez** [reply]
Oh wow, great!

I'd be very interested to know if you're able to differentiate between the _types_ of motion from the notification too, e.g. standard motion vs "person detected". That's one thing missing from the current polled ding event that I really hope to see.

**Reply by dgreif** [reply]
Here is a sample event:

```json
{
  "headers": {
    "Urgency": "high",
    "TTL": 1800
  },
  "notification": {
    "title": "Motion Alert",
    "body": "There is motion at your Testing",
    "timestamp": 1648852140
  },
  "data": {
    "ding": {
      "streaming_protocol": "ring_media_server",
      "device_name": "Testing",
      "doorbot_id": 111748380,
      "streaming_data_hash": "REDACTED",
      "device_kind": "cocoa_camera",
      "detection_type": "null",
      "id": "12345",
      "pod_id": 0,
      "location_id": "REDACTED",
      "image_uuid": "REDACTED",
      "properties": {
        "active_streaming_profile": "rms",
        "is_sidewalk": false
      }
    },
    "subtype": "motion",
    "action": "com.ring.push.HANDLE_NEW_motion"
  }
}
```

There are some really promising fields, like `image_uuid` which will solve a lot of the homebridge snapshot issues!  `detection_type` is likely to give you `person` or other values if you have smart notifications turned on, but I didn't get a chance to test that yet.

I'm out of time for today, but this still feels very promising.  I haven't been able to get the subscriptions registered from the Node side yet using `client_api/device`, but I'm hoping there is just something small that I'm overlooking.  Hopefully I get some time to work on it this weekend and get this nailed down.  Thanks again for all your input/help!

**Reply by sjonez** [reply]
This is very promising! Yes, I'm fairly certain you'll be able to see the different smart notifications - at least from the notification in my browser the body changes to "There is **a person**..." instead of "There is **motion**...". I assume it'll be the same for other types of smart notifications like "left packages" on the newer doorbells

**Reply by dgreif** [reply]
Success 😁.  Figuring out the senderId for the android app was the trickiest part, but I have my ways 😏.  I've got everything integrated and working reliably on my end.  Released version `10.0.0-beta.4` if you care to give it a try.  Thanks again for all your help and encouragement to push forward with this approach, I have been wanting to solve this problem for years!

**Reply by sjonez** [reply]
Just tested, it works great! Not only do we now get push notifications but I can now make use of smart notification types too, amazing!

Thanks for the super quick turnaround time from idea to code in just a few days :)

Just a quick question, there is still some polling required to get the camera status right? In order to retrieve battery level and siren/alarm state and such - I assume that polling hasn't been removed?

**Reply by dgreif** [reply]
Yep, that polling is still in place, defaulting to every 20 seconds. It can be increased with the settings if you're not using camera accessories for automations. I've been keeping a close eye on how Ring fetches that data and they still use polling from the web and the Android app. Hopefully someday they migrate that to websockets like the alarm system

**Reply by sjonez** [reply]
FYI I left the new example.ts running all night and while it is certainly picking up events I did notice this in the output:

```
(node:6801) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 ON_READY listeners added to [PushReceiver]. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
```
Any idea why I would have 11 listeners setup?

There were also quite a few of these, but not sure if that's just normal connections being dropped and re-established:

```
Error: read ECONNRESET
    at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20) {
  errno: -104,
  code: 'ECONNRESET',
  syscall: 'read'
}

```

**Reply by dgreif** [reply]
Thanks for pointing these out.  I think the MaxListeners warning is just going to happen because of how PushReceiver handles reconnects: https://github.com/Eneris/push-receiver/blob/master/src/client.ts#L159-L170.  It shouldn't be an issue.

The ECONNRESET are similar, it's just an error trying to read from the stream, likely due to a network blip of some kind.  I see a number of these in my logs as well. 

Fixing these would probably require a PR to Push Receiver.  I'll keep it in mind, but both seem harmless so far
